### ISSUE TITLE
unable to receive messages if swiped off from Recents

### ISSUE THREAD

**Issue by RamKromberg** [WhisperSystems/Signal-Android#4965]
Updated (2016-02-27):
    Version: 3.12
    Nexus is ok. x5pro still misbehaving. The bug is likely in the rom. But since there are so many affected devices and related regressions as system and apps are updated, I'll keep this issue open unless told otherwise.

2016-02-23:
    Version: 3.11.1
    Affected devices: Nexus 4 and Doogee X5 Pro. Stock unrooted roms or rooted roms regardless.
Possibly dupe of: https://github.com/WhisperSystems/Signal-Android/issues/1037

This used to only affect the x5pro but now the nexus recently shown the same bug. Also, Whatsapp doesn't display this bug so this is likely a Signal issue \ workaround-able android issue.

Previously:
Notes:
    Version: 3.8.1
    This is was originally thought to be an OEM related bug. Now it's been confirmed on two separate devices from two separate OEMs. One of which is running a custom ROM...
    Also, I have no idea how process and memory management actually works in Dalvik\ART so the terminology section is my ad-hoc attempt to work around this.

Terminology:
    Foregrounded: Visible. Full-screened. "The running" app.
    Backgrounded: Closed. Still visible in the Recents tasks list. Shown to be running under in Settings->Apps->Running Apps.
    Daemonized: Closed, Not visible in Recents. But still shown running as a service in Settings->Apps->Running Apps.
    Killed: Closed, Not visible in Recents. Not shown under Settings->Apps->Running Apps.

Scenario:
    2 clients; 1 Nexus 4 (works as expected), 1 Doogee X5Pro (misbehaving).

Working:
    1. When both phones have Signal foregrounded, messages pass back and forth smoothly and instantaneously.
    2. When Signal is Foregrounded on one phone while backgrounded on the other, a notification appears correctly.
    3. When the Nexus has Signal daemonized after being swiped off Recents, while the X5pro is sending through a foregrounded Signal, a notification appears correctly.
    4. When the X5pro has Signal backgrounded, notifications appear correctly while the Nexus is sending them through the foregrounded Signal client.
Not-working:
    Swiping off Signal from Recents doesn't daemonize the process. Instead it's killed and Signal messages fail to transfer. SMSs still work (I'm guessing a separate android daemon\service handles that).

Extra-credits:
    When remaining opened under Recents, Signal is indicated to occupy 26mb. When daemonized it's 6mb. Since I have no idea how Dalvik\Android actually manages memory, I can't say if that's address space (virtual memory) or resident memory or just what the Dalvik's process heap is pointing at...

TLDR: Seems on some phones, swiping off Signal from the Recents tasks-list doesn't daemonize properly. Instead, Signal is closed off entirely and Signal messages fail to transfer.

Edit:
    Updated reflecting further feedback. And version.

**Comment by xdattax** [comment]
same issue on lenovo zuk z1 cyanogenmod ,  even so when i reboot  phone i have to open signal to recieve messages 
i have afwall (signal ,google service) allowed

**Comment by xdattax** [comment]
https://gist.github.com/f0f29159a8bf627dc188

**Comment by RamKromberg** [comment]
Mine doesn't have root access enabled or installed. Just stock OEM. I actually considered rooting to install afwall and make sure only signal has access to ELEET and get rid of the bloatware in hopes some weird interaction is causing the problem... But if this happens in other phones using roms then it's probably just an unrelated bug (possible common driver code if not signal itself) so I'll hold off on unnecessary steps for now.

**Comment by tx3eh8IUD1** [comment]
I have the exact same issue on two phones.

Log 1: https://gist.github.com/anonymous/14ca27523c1bd5365ed3 
Log 2: https://gist.github.com/84d2a73657bd1cc36fd3

The first one is running stock rom from manufacturer (Huawei). 
The second (Galaxy S4, i9505, jfltexx) is running Cyanogenmod CM 13, nightly, 2016-02-08

On the i9505, **the issue did not appear on CM 12.1**. This leads me to believe that it is unrelated to signal but rather an OEM issue.

**Comment by 2-4601** [comment]
Related https://github.com/WhisperSystems/Signal-Android/issues/1037#issuecomment-36767446

**Comment by RamKromberg** [comment]
Update: Still bugged. Now also in the Nexus 4 stock rom. That is, swiping Signal off the Recents\Overview task manager in the nexus prevents incoming Signal messages. Using the latest Signal 3.11.1. Whatsapp works on the nexus (untested on the x5pro) so even if it's an Android bug, it's workaround-able. First post edited to reflect...

**Comment by 2-4601** [comment]
@RamKromberg That's strange, I'm unable to reproduce.

Tested with:
- Nexus 4, stock 5.1.1, Signal 3.12.0
- Huaweui U8800, 4.0.4, Signal 3.11.1

I get all messages, no difference if I swipe Signal from recents or not.

**Comment by agrajaghh** [comment]
Can't reproduce on:
- Nexus 4, Stock 5.1.1, Signal 3.12.0
- Sony Z3 Compact, Stock 5.1.1, Signal 3.12.0

**Comment by RamKromberg** [comment]
I can't triple check since I won't have access to the nexus until the weekend. Regardless, the nexus is running 3.11.1 while you're running 3.12 so I'm holding off self-flagellation until a 3.11.1 failure to reproduce is reported on the nexus ;)

**Comment by agrajaghh** [comment]
I doubt it's realted to 3.12.0, I think it tested this with older versions too..

**Comment by thorian93** [comment]
I have the exactly same issue.
Facts:
- Samsung Galaxy S5 Neo (SM-G903F)
- Android 5.1.1
- Signal 3.11.1

Has anybody found a workaround or is someone working on this?

**Comment by RamKromberg** [comment]
3.12 and\or a reboot sorted out the nexus. Signal still gets killed off when swiped-off overview in the x5pro but that's old news.

As far as workarounds go, my only reliable, albeit ad-hoc, workaround is more of an arrangement: Where by, a signal message is followed by a "." sms letting the recipient know he should click the notification to popup signal.
A less reliable solution is to test and see if Signal still receives messages if allowed to remain in the overview. I had _some_ luck with this but I also had enough unexplained failures that I'm not confident enough about this anymore.

**Comment by Trolldemorted** [comment]
@RamKromberg is this issue still present? Did you check if the signal process was definitely dead by running e.g. `ps | grep thought`?

**Comment by RamKromberg** [comment]
@Trolldemorted it still happens on the x5pro even with the recent factory android 6 firmware. The nexus died a while ago so I can't say regarding that...
As for the daemon, with the android5 firmwares the process was killed after a few hours of inactivity but with 6 it seems to persist indefinitely. However, there still occasions when notifications fail to appear until the client is opened.
Regardless, I'll close this issue. Seeing how dated the issue is and how it's becoming less and less frequent as updates are deployed, my guess is that it's a few different bugs that were, and are, being identified and fixed over time.
Thanks!

**Comment by loxK** [comment]
This issue still exists, I have hard times converting my contacts to Signal because of that issue.
